clang/include/clang/StaticAnalyzer/Core/PathSensitive/ExprEngine.h

```diff
@@ -323,12 +342,13 @@ class ExprEngine {
 
   /// ProcessBranch - Called by CoreEngine.  Used to generate successor
   ///  nodes by processing the 'effects' of a branch condition.
-  void processBranch(const Stmt *Condition,
-                     NodeBuilderContext& BuilderCtx,
-                     ExplodedNode *Pred,
-                     ExplodedNodeSet &Dst,
-                     const CFGBlock *DstT,
-                     const CFGBlock *DstF);
+  /// If the branch condition is a loop condition, IterationsFinishedInLoop is
+  /// the number of already finished iterations (0, 1, 2...); otherwise it's
+  /// std::nullopt.
```

Comment:
> Can't IterationsFinishedInLoop be a state trait? 

We could introduce it, but we'd need to maintain a nested stack of loops (1 iteration in that `while`, then 2 iterations in the `for`, then 1 iteration in this other `for` in the function called there...) to correctly monitor the number of iterations.

This would make the heuristic introduced in this commit slightly more accurate, but I think that for practical purposes the current code is already good enough, so I didn't invest time into implementing this loop iteration count stack. (However it would be a nice follow-up commit if I have time.)

Reply:
I would not say that `processBranch()` is part of the "engine API": it's a method which is called from one single location which is another closely related component of the engine, and I cannot imagine a valid use case for calling it from any other context. It's unfortunate that I'm extending this already long parameter list, but this kind of complexity cannot be entirely avoided in this sort of internal worker method.

I looked at the `ArrayInitLoopExpr` logic and it is not analogous to what I need. The main difference is that
- `IndexOfElementToConstruct` is used to __drive the iterations__ of the initialization loop, i.e. count the number of iterations and stop the loop when this reaches the `Size` of the array that's being constructed;
-  my code will a __small part of the body__ of larger loops which are driven by unrelated, external checks.

This difference means that in an array init loop the code _needs_ to manage an iteration count, while here I think it's much better to access the canonical `BlockCount` (which is, roughly speaking, the iteration count).

However, unfortunately the `BlockCount` is not the exactly right sort of iteration count for the heuristic that I want, because it's increasing monotonically with each evaluation of the loop condition, while I want to restart counting the number of iterations each time when we "reach the loop again" (due to a new iteration in an outer loop, or a new function call, or a goto that jumps to a place before the loop etc.).

Notice that `ArrayInitLoopExpr` does something similar when it adds a `LocationContext` to the key to distinguish initloops that happen within different function calls -- however, in my case something like this not be sufficient to restart iteration counting when we enter a new iteration of an outer loop.

This is the reason why I spoke about a "stack" in my previous comment -- to properly reset the iteration count each time we "reach the loop again", we'd need:
- a stack of the loops that we're inside of;
- tagging these loops with `LocationContext`s to distinguish executing the same code from different function calls;
- preferably some additional logic to handle `goto`, `break`, `return` and similar stuff.

In theory it's perfectly possible to implement this exact iteration counting, but I feel that in practice it's not worth the effort, because the `BlockCount` is a good enough approximation for this heuristic and it will make this (already complex) code significantly easier to maintain.

Reply:
Isn't binding the iteration count to the loop terminator sufficient?

Let's say we pass the terminator to `processBranch()` instead of the iteration count.
If there is no trait set for the terminator, we set it on the true branch, when the loop is entered.
If the terminator is already set, we know that we finished one iteration already and increment it.
If the condition is false and the loop is not entered again the trait is removed.

I'm not sure how well this would work though, since I didn't test it.